### ISSUE TITLE
State timeline panel: string based colors

### ISSUE THREAD

**Issue by PhilippDehner** [grafana/grafana#77421]
I want to see different colors based on the string values in my state timeline row. 

With the state timeline panel it is not working: 

![grafik](https://github.com/grafana/grafana/assets/49042626/0d3fbe4c-419c-4a07-9d80-6a06556283e5)

I can't see the changes between different states.

In the old discrete panel it is working:

![grafik](https://github.com/grafana/grafana/assets/49042626/ca762cf7-683b-465f-93b8-c6b249023186)

This feature is for all Grafana users.

**Comment by HbirdJ** [comment]
I solved this in grafana 10.2 by using a transform to convert the relevant field type to enum

![image](https://github.com/grafana/grafana/assets/13910021/b0cacd78-edca-46ca-babf-424344d4fc9d)

We just upgraded 10.4 and unfortunately this workaround is now far less effective. In 10.2 the enum values were always automatically mapped based on existing values. 10.3 or 10.4 added the ability to explicitly define the enum values, but this means I'd have to manually define all possible values which isn't an option in my use case.

Ideally I'd be able to tell the Transform to ALWAYS generate enum values from data.

**Comment by leeoniya** [comment]
> Ideally I'd be able to tell the Transform to ALWAYS generate enum values from data.

the problem with this is not having deterministic colors. if we just color the first seen value green, and next one red, then these can re-color on every data update, which would be terrifyingly bad. the only "auto" solution is to hash the strings and generate a color based on the hash, but this often ends up in pretty crappy palettes. thoughts?

**Comment by HbirdJ** [comment]
True, 'ideally' is probably the wrong phrasing on my end. While I do love a consistent color scheme, what matters more in the cases I run into is being able to quickly identify when state changed. 

If I needed a certain string value to always be the same color I'd use value mappings, but that's still only useful if you have a limited amount of potential string values.

I can see how re-coloring on data updates would be bad for some dashboards. I'm a frequent happy user of the color scheme option `Classic Palette (by series name)`, and before discovering my previous workaround was internally pining for something like `Classic Palette (by value)`.

**Comment by justinsmalley** [comment]
The "[Discrete](https://github.com/NatelEnergy/grafana-discrete-panel)" panel gets around this by adding a button called "+Add Current Values". The color mapping is non-deterministic in they way you describe, but with the click of a button will add the existing mappings based on your current view.
![image](https://github.com/grafana/grafana/assets/99200643/a9640bbf-ceb8-4082-8bbb-4d70c7893b0d)

**Comment by HbirdJ** [comment]
I've used that Discrete panel in the past for the reasons noted in this thread, but it seems like a dead project that's going to be a victim of the upcoming Angular depreciation.

**Comment by leeoniya** [comment]
> `Classic Palette (by value)`

this could be useful as an alternative to explicit value mappings. we'll give it some thought.

> The "[Discrete](https://github.com/NatelEnergy/grafana-discrete-panel)" panel gets around this by adding a button called "+Add Current Values"

this exists (kind of) in the type/Enum transformation shown above..to save you some manual entry. except we don't include colors there because it would basically replicate the same thing that Value Mappings already exists for (and with more flexibility there). this still doesnt solve the problem of auto-coloring unknown/not-yet-seen values.

> but it seems like a dead project that's going to be a victim of the upcoming Angular depreciation.

well, it's definitely dead; its author works at Grafana since 2019, and we wrote StateTimeline specifically because of the eventual Angular deprecation :D

**Comment by PhilippDehner** [comment]
@leeoniya The problem is not that the colors can change if I reload the data for a specific string value, but the problem is that the entire timeline is grey for each value, so I cannot differentiate between the different values

**Comment by mrgiraldoo** [comment]
Hello, is there any progress on this?

**Comment by redbaron** [comment]
@HbirdJ , many thanks for "Convert to Enum" trick, works wonders I wouldn't think about solving it this way without your tip.

**Comment by llamafilm** [comment]
This is still needed! We have many state timelines where I don't know all the possible values in advance. So I just want it to use random colors to make it easy to see when states change.

**Comment by pfVer** [comment]
A workaround I am using is to set the line thickness to a non-zero value. That does not give you different colors, but still clear boundaries between values (see screenshot below):

<img width="1920" height="445" alt="Image" src="https://github.com/user-attachments/assets/63cb6435-ed2a-4851-a06f-7a84a2e7cf37" />
